### ISSUE TITLE
PackedStringArray.to_byte_array creates an array of pointers, not values

### ISSUE THREAD

**Issue by gormster** [godotengine/godot#66526]
### Godot version

4.0.beta1

### System information

macOS 12.5.1 arm64

### Issue description

Calling `to_byte_array` on `PackedStringArray` produces a byte representation of the internal pointers used to manage the strings in PackedStringArray, rather than the string data itself.

This is… interesting, as it demonstrates that `PackedStringArray` isn't really a "packed string array" at all. It's not a packed array of strings, it's a regular old array of pointers that can themselves be pointing anywhere in memory.

I've been working on a patch that will allow easy conversion to and from a true packed array of strings (with various encodings), separated by NULLs, but the more I work on it, the more I think that this isn't a good way of doing things. For one thing, Godot has a [canonical way of encoding variable-length data](https://github.com/godotengine/godot/blob/92bcd3c01d5188480793c03b2b50e97363ceb624/core/io/marshalls.h#L215), and this isn't it.

`PackedStringArray` seems like a bit of an odd duck among the `PackedXArray` classes. It's the only one that stores variable-length data, and the only one that stores it out-of-line. It feels like most of the time it should probably be a `TypedArray<String>` instead, or a plain old `Vector<String>` when not exposed to the API directly. But it's probably too late to make that change now.

At the very least, `PackedStringArray.to_byte_array` should not be defined in the API.

### Steps to reproduce

1. Create a GDScript file.
2. Create a PackedStringArray, populated with some values.
3. Call `to_byte_array` on that array.
4. Print the result.

You will see the result has exactly eight values for each string, containing apparently random bytes, instead of the bytes representing the characters in the strings.

### Minimal reproduction project

[PackedStringArrayBugDemo.zip](https://github.com/godotengine/godot/files/9662309/PackedStringArrayBugDemo.zip)

**Comment by Calinou** [comment]
Can you reproduce this on 3.5?

**Comment by timothyqiu** [comment]
`PoolStringArray` does not have any method similar to `to_byte_array()`.

Also, I think the "good" version in the MRP is not the expected result either. `var_to_bytes()` uses variant encoding which prepends extra infomation.

**Comment by gormster** [comment]
> Also, I think the "good" version in the MRP is not the expected result either. `var_to_bytes()` uses variant encoding which prepends extra infomation.

yeah, I mean, I was being a bit liberal with "good" and "bad" there. but hey, two hard problems, right?

That said, I'm struggling to think of a serialisation situation where this actually matters? It's eight extraneous bytes. If you really care, just slice them off. If you're in a situation where you need a packed string array to be in a particular format because you don't control the deserialisation mechanism, you'll probably need to encode them one-by-one anyway; what are the chances that whatever serialisation we bless with `to_byte_array` will be the correct one for your use case?

**Comment by OsakiTsukiko** [comment]
Yea, I think it's best the method is removed, as it only causes confusion and is made even worse by the docs stating: `Returns a PackedByteArray with each string encoded as bytes.`. I don't think there's an actual use for this function.

**Comment by geekley** [comment]
I was about to ask about this, then I found this issue.

The docs should be more clear about the internal structure of both types. It should clearly answer:
- Is every `String` immutable? If so, this isn't obvious in the explanation from the docs, from the way it's phrased (the "modification" returning a new String doesn't automatically imply the original is unmodified):
    > Strings are reference-counted and use a copy-on-write approach (every modification to a string returns a new String), so passing them around is cheap in resources.
    
    Is the hash calculated only upon request? Is it stored with the string?
    Also how many bytes per character? Is it a fixed amount? Is the in-memory `String` UTF16 or UTF8 encoded, like in other languages? Do characters with more than 16 bits occupy 2 chars or just one?

- What's the inner structure of a `PackedStringArray` in memory? How does it differ from `Array[String]`?
    > Packs data tightly, so it saves memory for large array sizes.
    
    The description above is not enough, as it won't clearly explain whether this is:
    - array of pointers to regular `String` values, with no copy of characters
    - array of lengths followed by char data for every item (where char data is copied when assigned from a `String`)
    - some other weird structure, e.g. array of (length + chars) values

This info is important and needs to be in the documentation. Specially for implementation decisions where you need to have certain efficiency guarantees.

**Comment by Splizard** [comment]
@AThousandShips has indicated that the function should be deprecated.
